web_src/js/modules/fetch.js

```diff
@@ -0,0 +1,31 @@
+const {csrfToken} = window.config;
+
+function request(url, {headers, json, ...other} = {}) {
+  return window.fetch(url, {
+    headers: {
+      'x-csrf-token': csrfToken,
+      ...(json !== undefined && {'content-type': 'application/json'}),
+      ...headers,
+    },
+    ...(json !== undefined && {body: JSON.stringify(json)}),
+    ...other,
+  });
+}
+
+export const GET = (url, opts) => request(url, {method: 'GET', ...opts});
+export const POST = (url, opts) => request(url, {method: 'POST', ...opts});
+export const PATCH = (url, opts) => request(url, {method: 'PATCH', ...opts});
+export const PUT = (url, opts) => request(url, {method: 'PUT', ...opts});
+export const DELETE = (url, opts) => request(url, {method: 'DELETE', ...opts});
+
+// network errors are currently only detectable by error message
+// based on https://github.com/sindresorhus/p-retry/blob/main/index.js
+const networkErrorMsgs = new Set([
+  'Failed to fetch', // Chrome
+  'NetworkError when attempting to fetch resource.', // Firefox
+  'The Internet connection appears to be offline.', // Safari
+]);
```

Comment:
I think it doesn't need this trick.

Using `pagehide` event listener could know the page is unloading ahead.

Reply:
I tried that earlier but it seems timing-wise the error is much sooner:

<img width="736" alt="image" src="https://github.com/go-gitea/gitea/assets/115237/8aead0f7-5cb5-4f9f-b1cc-4e491128aed6">

Maybe I'm misunderstanding the problem and the error is not actually cause by fetch aborting because of unload but something else.

Reply:
Hmm ... maybe different browsers have different behaviors?

The `pagehide` is fired before the fetch aborting in Chrome.

I will take a look in Firefox.

Reply:
Timing should be the same in all browsers. I think the problem with the error might be because of vue unmounting maybe?

Reply:
According to my test, not the same.

* In Chrome, the "pagehide" comes before "fetch TypeError"
* But in Firefox, the "pagehide" comes after "fetch TypeError" (eg: at least a 20-30ms delay)

According to MDN: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypeError


I think for this fetch error, it's safe to just check whether e is TypeError. If yes, only log it in console.

Reply:
ps: Firefox does very bad for this case, it doesn't distinguish the aborting error and other errors. Fetching an invalid URL also causes `NetworkError when attempting to fetch resource.` So the "fetch" functions will lost the ability to report network error to end users (I am open for this problem)

Reply:
TypeError is a very generic error class in JS, not specific to fetch at all. But we could just wrap the fetch in another fetch clause for it I guess to avoid  those string checks.

Reply:
Actually, I do see 500 errors in the server logs for this as well, so it might also just be a backend bug, thought those responses never show a 500 status code on frontend:

```
2023/09/09 14:04:38 ...eb/routing/logger.go:102:func1() [I] router: completed POST /silverwind/symlink-test/actions/runs/33/jobs/0 for 127.0.0.1:63901, 500 Internal Server Error in 53.2ms @ actions/view.go:127(actions.ViewPost)

Reply:
Yup, I also agree to wrap the fetch in a catch, then no need to do `isNetworkError` check for every usage.

Reply:
> Actually, I do see 500 errors in the server logs for this as well, so it might just be a backend bug after all, thought those responses never show a 500 status code on frontend:

No, I think they are different. In many cases, backend 500 error is reported incorrectly and the frontend do not see the response.

And, a correct 500 response doesn't cause `fetch` to throw an error.

Reply:
I tried wrapping the fetch specifically and check for `TypeError`, but in my testing, some errors slipped through that for some reason. I guess we should first investigate this backend error before doing more on frontend.

Reply:
> I tried wrapping the fetch specifically and check for `TypeError`, but in my testing, some errors slipped through that for some reason. I guess we should first investigate this backend error before doing more on frontend.

I didn't get your meaning. I think there are only a few cases:

1. If `fetch` throws exception, it is a network error.
2. If `fetch` response code is >=500, it is server internal error, it could show a general error message to end users
3. Otherwise, return the response, let the caller handle it.


Case 1 and Case 2 can be wrapped in our general `request` function.

Reply:
Ah, I was under the impression fetch can throw on non-network errors too, but it seems you are correct and any thrown error is a network error, which at most we could log to the console or show a toast depending on use case.

Reply:
> > In many cases, backend 500 error is reported incorrectly and the frontend do not see the response.
> 
> You think these 500s are not really ones? Why does the logging not show a error reason in backend log?

Because if the browser aborts the connection, then backend "ctx" is cancelled, then the backend code could fail at any point.

Reply:
Right. Seem like if the backend can detect client cancellation by socket closure, it should not log as a 500 because that 500 was never delivered.

Reply:
Should be solved in https://github.com/go-gitea/gitea/pull/26985/commits/9f91c9e4cbad19e670e532da18dba483bde2d430. I had missed that there were two fetches, wrapping both in the `try` made it work.